docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+The default `spec.user` can be "reconciler".
+A user field will be added to the Kustomization and HelmRelease spec.
+It enables a partial override for the namespaced username the controller uses for that particular `Kustomization`/`HelmRelease`.
+Namespace Owners can use the user override to optionally specify flux `Users` with different `Roles`/`RoleBindings` for managing each workload,
+similar to Pods specifying non-default serviceAccounts from the same Namespace.
+- **Con**: RoleBinding against a FluxUser requires specifying the namespace in the username string which can be hard to do in a generic way with kustomize -- would need to document a kustomize example
+
+Flux controllers will also impersonate two Groups.
+The first is `flux:users` which is a generic group, analogous to kubernetes' `system:authenticated` -- binding against this group allows all reconcilers to access particular resources via RBAC.
+The second is `flux:users:{{namespace}}`, a namespace specific group which enables Namespace Owners to rolebind for reconcilers from another Namespace without mandating a particular username.
+Neither of these groups are optional or overrideable -- they are a given property of each reconciler's Namespace and identity.
+Note `users` is plural in the Group names, but not for a single User -- this matches kubernetes SA's.
+It could cause typos, but not matching k8s plural semantics could also be confusing. We could be silly and add both without notable consequence. 🚲🏠
+
+Impersonation of Users and Groups is done via API Headers that are permitted for the controller ServiceAccounts.
+- **Con**: controller SA's can impersonate any user or group, not just flux:user namespaced ones is explicitly specified via a ClusterRole (impractical,bad-ux)
+       controller code needs to be trusted not to impersonate 
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: flux-user-group-impersonator
+rules:
+- apiGroups: [""]
+  resources: ["users", "groups"]
+  verbs: ["impersonate"]
+```
+
+Using Namespaced Users and Groups provides a security boundary in the same way the ServiceAccount User names do.
+It's never allowed for a reconcile to impersonate a non-namespaced User or user from outside the Namespace of the particular reconcile resource.
+If a reconcile needs access to resources from another Namespace, the other Namespace can have a RoleBinding targeting the reconciler's differently namespaced User name.
+Alternatively, a ClusterRoleBinding on the reconciler's namespaced User name can grant access to all Namespaces.
+
+#### serviceAccount impersonation
+`spec.serviceAccountName` will remain supported, but impersonating Users is preferable because it reduces the attack surface within a Namespace.
```

Comment:
Could it be an option to remove it?

If the controller starts up without impersonation permissions, it can fall back to using the pod service account ("namespaced mode" or cluster admin, up to how the SA permissions are configured).

Reply:
This is not an option IMO, many users are running Flux in production with this setting, not to mention that flux1 multi- tenancy was based on SAs, we’ll be cutting off all these users which I find unacceptable.

Reply:
@stefanprodan @stealthybox a lot of the changes will already need a migration path - if this makes the product easier and more secure, wouldn't it be worth it? We won't be able to do it after the fact.

Do we see a reason maintaining two ways of doing multi-tenancy? Seems like a lot of work keeping track of it just because of compatibility.

Reply:
I think there should be a deprecation path but not removed immediately. Removing serviceaccountname immediately will make it difficult for existing users to migrate. One solution could be to look into adding some deprecation notice to convince users to migrate over to users.